File: src/tracks/shadow/restore.ts
import { execFileSync } from "node:child_process";
import * as fs from "node:fs";
import * as path from "node:path";
import { parseFlag } from "../../args";
import {
  addToManifest,
  defaultManifestPath,
  MANIFEST_FILENAME,
  readManifest,
} from "../../manifest";
import { resolveRepoRoot } from "../../repo-root";
import { safeRemove, safeWriteFile } from "../../safe-write";
import { resolveProjectId, resolveShadowRepoPath } from "./paths";

/**
 * `omc-sync restore --track shadow [--ref <sha-or-ref>]`
 *
 * Implements Part B, Architecture step 6 of
 * .omc/plans/shadow-ref-git-sync-for-omc-artifacts.md (US-007 / AC-B2,
 * AC-B3): materializes the tree at `refs/omc/<project-id>/data` (or a
 * `--ref` override) back onto disk at `.omc/<path>` in the anchor repo.
 *
 * This is a true tree-sync, not an additive overlay:
 *   - every path present in the target tree is (re)written from the
 *     blob's exact bytes;
 *   - every path currently listed in the manifest that is *not* present in
 *     the target tree, but *was* present at some earlier commit reachable
 *     from the target ref (i.e. it was genuinely synced once and is now
 *     genuinely gone), is deleted, if present on disk.
 *
 * Deletion is deliberately scoped to (ref history \ target tree), never to
 * (current manifest \ target tree): a path can be manifest-listed and
 * absent from the target tree merely because it was never successfully
 * synced yet (e.g. it has always matched the advisory scan in `push`, so no
 * commit in this ref's history ever contained it). Deleting the local file
 * in that case would destroy content that was never backed up. Checking
 * the ref's own history (via `git log -- <path>`) is what distinguishes a
 * genuine prior-then-gone deletion from a path that simply never made it
 * into the shadow ref in the first place.
 *
 * Scope is otherwise limited to the manifest ∪ the target tree — this never
 * touches files under `.omc/` that this tool has no knowledge of.
 */
export function run(args: string[]): void {
  const { value: refFlag } = parseFlag(args, "ref");
  const repoRoot = resolveRepoRoot();
  const projectId = resolveProjectId(repoRoot);
  const shadowRepoPath = resolveShadowRepoPath(projectId);

  if (!fs.existsSync(shadowRepoPath)) {
    throw new Error(
      `restore --track shadow: no shadow repo found at ${shadowRepoPath} — run \`omc-sync init --track shadow\` first`,
    );
  }

  const gitDir = `--git-dir=${shadowRepoPath}`;
  const refName = refFlag ?? `refs/omc/${projectId}/data`;

  // On a fresh machine (a shadow repo that was just `init`-ed but never
  // pushed from), the local ref doesn't exist yet — only `origin` knows
  // about it. Best-effort fetch it into the matching local ref name before
  // reading the tree; if this fails (offline, ref never pushed, or an
  // explicit --ref that isn't a remote-tracking ref name), fall through to
  // `listTree`, which surfaces a clear error if the ref is unresolvable
  // both locally and remotely.
  if (!refFlag) {
    tryFetchRef(gitDir, refName);
  }

  const targetPaths = listTree(gitDir, refName);
  const targetSet = new Set(targetPaths);
  const omcRoot = path.join(repoRoot, ".omc");

  for (const relPath of targetPaths) {
    if (relPath === MANIFEST_FILENAME) {
      // The manifest itself is handled specially below via a union-merge
      // into the LOCAL manifest, never wholesale-overwritten from the
      // incoming tree like an ordinary file — see mergeIncomingManifest.
      continue;
    }
    const content = readBlob(gitDir, refName, relPath);
    const destPath = path.join(repoRoot, ".omc", relPath);
    safeWriteFile(omcRoot, destPath, content);
  }

  const localManifestPath = defaultManifestPath(repoRoot);
  const manifestPaths = readManifest(localManifestPath);
  for (const relPath of manifestPaths) {
    if (targetSet.has(relPath)) {
      continue;
    }
    if (!everSyncedInHistory(gitDir, refName, relPath)) {
      // This path is absent from the target tree, but the shadow ref's own
      // history shows it was never actually synced (e.g. it has always
      // matched the advisory scan). Its absence carries no deletion
      // intent, so any local copy is left untouched.
      continue;
    }
    const destPath = path.join(repoRoot, ".omc", relPath);
    safeRemove(omcRoot, destPath);
  }

  if (targetSet.has(MANIFEST_FILENAME)) {
    mergeIncomingManifest(gitDir, refName, localManifestPath);
  }
}

/**
 * Parses the incoming manifest blob's raw lines (same skip-blank/skip-
 * comment rules as `readManifest`) and UNION-merges each valid line into the
 * local manifest via `addToManifest` — additive only, so a pre-existing
 * local-only entry the incoming manifest doesn't mention is never removed or
 * overwritten. Reads via `readBlob` (git's own object store), so there's no
 * filesystem symlink-escape surface to guard against here.
 */
function mergeIncomingManifest(
  gitDir: string,
  refName: string,
  localManifestPath: string,
): void {
  const raw = readBlob(gitDir, refName, MANIFEST_FILENAME).toString("utf8");
  const lines = raw
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0 && !line.startsWith("#"));

  for (const line of lines) {
    try {
      addToManifest(localManifestPath, line);
    } catch (err) {
      // A hand-edited or otherwise malformed incoming manifest could contain
      // an out-of-bounds entry (absolute path / `../` traversal);
      // addToManifest fails closed on those. Skip just that one line with a
      // warning rather than aborting the whole restore.
      process.stderr.write(
        `omc-sync: skipping invalid incoming manifest entry '${line}': ${(err as Error).message}\n`,
      );
    }
  }
}

/**
 * Returns true if `relPath` was ever added/modified/removed in some commit
 * reachable from `refName` — i.e. it was genuinely synced into the shadow
 * ref's history at some point, regardless of whether it's present in the
 * current tip's tree. Used to distinguish a genuine (previously-synced,
 * now-deleted) path from one that simply never made it into any commit.
 */
function everSyncedInHistory(gitDir: string, refName: string, relPath: string): boolean {
  try {
    const out = execFileSync(
      "git",
      [gitDir, "log", "--format=%H", "-1", refName, "--", relPath],
      { encoding: "utf8", stdio: ["ignore", "pipe", "ignore"] },
    ).trim();
    return out.length > 0;
  } catch {
    return false;
  }
}

function tryFetchRef(gitDir: string, refName: string): void {
  try {
    execFileSync("git", [gitDir, "fetch", "origin", `+${refName}:${refName}`], {
      stdio: ["ignore", "pipe", "pipe"],
    });
  } catch {
    // Best-effort: no origin, offline, or nothing has ever been pushed.
    // `listTree` below will surface a clear error if the ref truly can't
    // be resolved locally either.
  }
}

function listTree(gitDir: string, refName: string): string[] {
  let out: string;
  try {
    out = execFileSync("git", [gitDir, "ls-tree", "-r", "--name-only", refName], {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"],
    });
  } catch (err) {
    const stderr = (err as { stderr?: Buffer | string }).stderr;
    const detail = stderr ? stderr.toString() : (err as Error).message;
    throw new Error(
      `restore --track shadow: failed to read ref '${refName}': ${detail}`,
    );
  }
  return out
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
}

/**
 * Extracts a blob's exact bytes as a `Buffer` (not a string round-trip), so
 * CRLF (or any other byte sequence) content round-trips byte-for-byte.
 */
function readBlob(gitDir: string, refName: string, relPath: string): Buffer {
  return execFileSync("git", [gitDir, "show", `${refName}:${relPath}`], {
    maxBuffer: 100 * 1024 * 1024,
  });
}

File: go/internal/tracks/shadow/restore.go
package shadow

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"plan-sync/go/internal/args"
	"plan-sync/go/internal/manifest"
	"plan-sync/go/internal/reporoot"
	"plan-sync/go/internal/root"
	"plan-sync/go/internal/safewrite"
	"plan-sync/go/internal/shadowpaths"
)

// Restore is the Go port of src/tracks/shadow/restore.ts's `run`, reached
// from the CLI as `plan-sync pull --track shadow [--ref <sha-or-ref>]
// [--root <dir>]` (src/commands/pull.ts dispatches the shadow track
// straight into this module; the standalone `restore` command was unified
// into `pull`).
//
// It materializes the tree at `refs/plan-sync/<project-id>/<root>/data`
// (or a `--ref` override) back onto disk at `<rootDir>/<path>` in the
// anchor repo.
//
// This is a true tree-sync, not an additive overlay:
//   - every path present in the target tree is (re)written from the blob's
//     exact bytes;
//   - every path currently listed in the manifest that is *not* present in
//     the target tree, but *was* present at some earlier commit reachable
//     from the target ref (i.e. it was genuinely synced once and is now
//     genuinely gone), is deleted, if present on disk.
//
// Deletion is deliberately scoped to (ref history \ target tree), NEVER to
// (current manifest \ target tree): a path can be manifest-listed and
// absent from the target tree merely because it was never successfully
// synced yet (e.g. it has always matched `push`'s advisory secret-shape
// scan, so no commit in this ref's history ever contained it). Deleting the
// local file in that case would destroy content that was never backed up.
// Checking the ref's own history (via `git log -- <path>`, see
// everSyncedInHistory) is what distinguishes a genuine
// previously-synced-then-removed path from one that simply never made it
// into the shadow ref in the first place. Do not "simplify" this into a
// manifest-minus-tree set difference — that is the exact regression this
// scoping exists to prevent (docs/HARDENING-HISTORY.md, round 1 finding 1
// chained with round 3 findings 9 and 10).
//
// Scope is otherwise limited to the manifest ∪ the target tree — this never
// touches files under `<rootDir>/` that this tool has no knowledge of.
//
// Every write goes through safewrite.SafeWriteFile and every delete through
// safewrite.SafeRemove; there is no other way this function touches a
// destination path.
//
// Failure behavior (Phase-1 observability requirement): an unresolvable
// ref — a corrupt/missing local ref with an unreachable or absent origin —
// returns a non-nil error with a non-empty message BEFORE any destination
// is written or removed, so the process exits non-zero with a real
// diagnostic on stderr and every already-on-disk local file is left exactly
// as it was. The ref is fully enumerated (listTree) up front, ahead of the
// first mutation, precisely so a bad ref can never produce a partial write.
func Restore(argv []string) error {
	refFlag, rest1 := args.ParseFlag(argv, "ref")
	rootFlag, _ := args.ParseFlag(rest1, "root")

	repoRoot, err := reporoot.ResolveRepoRoot()
	if err != nil {
		return err
	}
	rootDir, err := root.ResolveRootDir(repoRoot, rootFlag)
	if err != nil {
		return err
	}
	projectID, err := shadowpaths.ResolveProjectId(repoRoot)
	if err != nil {
		return err
	}
	shadowRepoPath, err := shadowpaths.ResolveShadowRepoPath(projectID, rootDir)
	if err != nil {
		return err
	}

	// os.Lstat, not os.Stat: a symlink planted at the shadow repo path
	// should be reported as present-but-whatever-it-is rather than silently
	// resolving through to somewhere else, and a dangling one must not read
	// as "not initialized" and trigger a confusing suggestion to re-init.
	if _, err := os.Lstat(shadowRepoPath); err != nil {
		return fmt.Errorf(
			"restore --track shadow: no shadow repo found at %s — run `plan-sync init --track shadow` first",
			shadowRepoPath)
	}

	gitDir := gitDirFlag(shadowRepoPath)
	refName := refFlag
	if refName == "" {
		refName, err = shadowpaths.ResolveShadowRefName(projectID, rootDir)
		if err != nil {
			return err
		}
		// On a fresh machine (a shadow repo that was just init-ed but never
		// pushed from), the local ref doesn't exist yet — only `origin`
		// knows about it. Best-effort fetch it into the matching local ref
		// name before reading the tree; if this fails (offline, ref never
		// pushed, no origin), fall through to listTree, which surfaces a
		// clear error if the ref truly can't be resolved locally either.
		//
		// Only for the derived ref: an explicit --ref may well be a raw sha
		// or some other name that isn't a remote-tracking ref, matching the
		// TypeScript original's `if (!refFlag)` guard.
		tryFetchRef(gitDir, refName)
	}

	targetPaths, err := listTree(gitDir, refName)
	if err != nil {
		return err
	}
	targetSet := make(map[string]bool, len(targetPaths))
	for _, p := range targetPaths {
		targetSet[p] = true
	}

	omcRoot := filepath.Join(repoRoot, rootDir)

	for _, relPath := range targetPaths {
		if relPath == manifest.ManifestFilename {
			// The manifest itself is handled specially below via a
			// union-merge into the LOCAL manifest, never wholesale-
			// overwritten from the incoming tree like an ordinary file —
			// see mergeIncomingManifest.
			continue
		}
		content, err := readBlob(gitDir, refName, relPath)
		if err != nil {
			return fmt.Errorf("restore --track shadow: failed to read '%s' at ref '%s': %w",
				relPath, refName, err)
		}
		safewrite.SafeWriteFile(omcRoot, filepath.Join(omcRoot, relPath), content)
	}

	localManifestPath := filepath.Join(omcRoot, manifest.ManifestFilename)
	for _, relPath := range manifest.ReadManifest(localManifestPath) {
		if targetSet[relPath] {
			continue
		}
		if !everSyncedInHistory(gitDir, refName, relPath) {
			// This path is absent from the target tree, but the shadow
			// ref's own history shows it was never actually synced (e.g. it
			// has always matched the advisory scan). Its absence carries no
			// deletion intent, so any local copy is left untouched.
			continue
		}
		safewrite.SafeRemove(omcRoot, filepath.Join(omcRoot, relPath))
	}

	if targetSet[manifest.ManifestFilename] {
		if err := mergeIncomingManifest(gitDir, refName, localManifestPath); err != nil {
			return err
		}
	}
	return nil
}

// mergeIncomingManifest parses the incoming manifest blob's raw lines (same
// skip-blank/skip-comment rules as manifest.ReadManifest) and UNION-merges
// each valid line into the local manifest via manifest.AddToManifest —
// additive only, so a pre-existing local-only entry the incoming manifest
// doesn't mention is never removed or overwritten. Reads via readBlob
// (git's own object store), so there is no filesystem symlink-escape
// surface to guard against here.
func mergeIncomingManifest(gitDir, refName, localManifestPath string) error {
	raw, err := readBlob(gitDir, refName, manifest.ManifestFilename)
	if err != nil {
		return fmt.Errorf("restore --track shadow: failed to read the incoming manifest at ref '%s': %w",
			refName, err)
	}

	for _, line := range strings.Split(string(raw), "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if err := manifest.AddToManifest(localManifestPath, line); err != nil {
			// A hand-edited or otherwise malformed incoming manifest could
			// contain an out-of-bounds entry (absolute path / `../`
			// traversal); AddToManifest fails closed on those. Skip just
			// that one line with a warning rather than aborting the whole
			// restore.
			fmt.Fprintf(os.Stderr,
				"plan-sync: skipping invalid incoming manifest entry '%s': %v\n", line, err)
		}
	}
	return nil
}

// everSyncedInHistory reports whether relPath was ever added/modified/
// removed in some commit reachable from refName — i.e. it was genuinely
// synced into the shadow ref's history at some point, regardless of whether
// it is present in the current tip's tree. This is what distinguishes a
// genuine (previously-synced, now-deleted) path, whose local copy restore
// should remove, from one that simply never made it into any commit, whose
// local copy restore must leave alone.
//
// Any failure (unresolvable ref, git error) reports false — "no evidence
// this was ever synced" — which fails toward NOT deleting. That direction
// is the safe one: the cost of a false negative is a stale local file, the
// cost of a false positive is destroyed, never-backed-up user content.
func everSyncedInHistory(gitDir, refName, relPath string) bool {
	out, err := runGit(gitDir, "log", "--format=%H", "-1", refName, "--", relPath)
	if err != nil {
		return false
	}
	return strings.TrimSpace(out) != ""
}

// tryFetchRef best-effort fetches refName from origin into the identically
// named local ref. Failures (no origin, offline, nothing ever pushed) are
// deliberately swallowed: listTree runs next and surfaces a clear error if
// the ref truly cannot be resolved locally either, so failing here would
// only replace a precise diagnostic with a vaguer one — and would break the
// ordinary offline case where the local ref is already correct.
func tryFetchRef(gitDir, refName string) {
	_, _ = runGit(gitDir, "fetch", "origin", "+"+refName+":"+refName)
}

// listTree enumerates every path in the tree at refName. It is called
// before any destination is touched, so an unresolvable ref aborts the
// whole restore with a clear error and zero writes.
func listTree(gitDir, refName string) ([]string, error) {
	out, err := runGit(gitDir, "ls-tree", "-r", "--name-only", refName)
	if err != nil {
		return nil, fmt.Errorf("restore --track shadow: failed to read ref '%s': %w", refName, err)
	}

	paths := []string{}
	for _, line := range strings.Split(out, "\n") {
		if trimmed := strings.TrimSpace(line); trimmed != "" {
			paths = append(paths, trimmed)
		}
	}
	return paths, nil
}

// readBlob extracts a blob's exact bytes (not a string round-trip), so CRLF
// — or any other byte sequence — round-trips byte-for-byte. Combined with
// the shadow repo's `core.autocrlf=false` and `* -text` attributes pinned
// by Init, this is what makes the push/restore round-trip byte-identical.
//
// Bounded at maxBlobBytes (mirroring src/tracks/shadow/restore.ts:204's
// maxBuffer), so an oversized blob fails cleanly here — before any of
// Restore's target paths are written — rather than buffering unboundedly
// and risking an OOM mid-loop after some paths are already written.
func readBlob(gitDir, refName, relPath string) ([]byte, error) {
	return runGitBytesBounded(maxBlobBytes, gitDir, "show", refName+":"+relPath)
}
